### PR TITLE
MappedEditsProvider API 

### PR DESCRIPTION
**Pull request by ulugbekna**

### CODE REVIEW

File: src/vs/workbench/contrib/mappedEdits/common/mappedEdits.contribution.ts [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+
+import { CancellationTokenSource } from 'vs/base/common/cancellation';
+import { URI } from 'vs/base/common/uri';
+import { ITextModelService } from 'vs/editor/common/services/resolverService';
+import * as nls from 'vs/nls';
+import { Action2, registerAction2 } from 'vs/platform/actions/common/actions';
+import { ServicesAccessor } from 'vs/platform/instantiation/common/instantiation';
+import { IMappedEditsService, MappedEditsContext } from 'vs/workbench/services/mappedEdits/common/mappedEdits';
+
+class ExecuteMappedEditsProvider extends Action2 {
```
Comment (aeschli) [inline src/vs/workbench/contrib/mappedEdits/common/mappedEdits.contribution.ts:14]: does this need to be an Action? I think registering a command is enough

`CommandsRegistry.registerCommand('_executeMappedEditsProvider', ...)`

File: src/vs/workbench/services/mappedEdits/browser/mappedEditsService.ts [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+import { ITextModel } from 'vs/editor/common/model';
+import { IMappedEditsProvider, IMappedEditsService, MappedEditsContext } from 'vs/workbench/services/mappedEdits/common/mappedEdits';
+import { score } from 'vs/editor/common/languageSelector';
+import { WorkspaceEdit } from 'vs/editor/common/languages';
+import { InstantiationType, registerSingleton } from 'vs/platform/instantiation/common/extensions';
+
+export class MappedEditsService implements IMappedEditsService {
+	readonly _serviceBrand: undefined;
+
+	private readonly _providers = new Set<IMappedEditsProvider>();
```
Comment (aeschli) [inline src/vs/workbench/services/mappedEdits/browser/mappedEditsService.ts:16]: There's the `LanguageFeatureRegistry` that you can use

**Review by aeschli — DISMISSED** [review]

File: src/vs/workbench/api/common/extHost.protocol.ts [hunk 3]
```diff
@@ -371,6 +371,16 @@ export interface IShareableItemDto {
 	selection?: IRange;
 }
 
+export interface IRelatedContextItemDto {
+	readonly uri: UriComponents;
+	readonly range: IRange;
+}
+
+export interface IMappedEditsContextDto {
+	selections: ISelection[]; //FIXME@ulugbekna: is this serializable? should I use ISelection?
```
Comment (alexdima) [inline src/vs/workbench/api/common/extHost.protocol.ts:380]: Yes, using `ISelection` is OK.

File: src/vs/workbench/api/common/extHost.protocol.ts [hunk 4]
```diff
@@ -1315,6 +1325,11 @@ export interface MainThreadShareShape extends IDisposable {
 	$unregisterShareProvider(handle: number): void;
 }
 
+export interface MainThreadMappedEditsShape extends IDisposable {
```
Comment (alexdima) [inline src/vs/workbench/api/common/extHost.protocol.ts:1328]: Given this is about language servers / language smarts, I suggest moving these into `MainThreadLanguageFeaturesShape` / `ExtHostLanguageFeaturesShape` like all the other language smarts

File: src/vs/workbench/services/mappedEdits/common/mappedEdits.ts [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+		codeBlocks: string[],
+		context: MappedEditsContext,
+		token: CancellationToken
+	): Promise<WorkspaceEdit | null>;
+}
+
+
+export const IMappedEditsService = createDecorator<IMappedEditsService>('mappedEditsService');
+
+export interface IMappedEditsService {
```
Comment (alexdima) [inline src/vs/workbench/services/mappedEdits/common/mappedEdits.ts:57]: I suggest merging this into `ILanguageFeaturesService` and using patterns similar from there.

File: src/vs/workbench/services/mappedEdits/browser/mappedEditsService.ts (hunk 2 above)
Comment (alexdima) [reply src/vs/workbench/services/mappedEdits/browser/mappedEditsService.ts:16]: Yes! That allows to score providers based on their selector specificity!

**Review by aeschli — DISMISSED** [review]

**Review by alexdima — APPROVED** [review]
👏 Very nice!

**Pull request merged**